File: languages/montithings/src/main/java/montithings/trafos/patterns/LinearRegression.java
package montithings.trafos.patterns;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public abstract class LinearRegression {

    protected Double predictValue(Double input, List<Double> values) {
        List<Double> xValues = this.xValues(values);

        Double meanX = this.mean(xValues);
        Double meanY = this.mean(values);

        Double stdX = this.std(xValues, meanX);
        Double stdY = this.std(values, meanY);

        Double covariance = this.covariance(xValues, values, meanX, meanY);

        Double slope = this.slope(covariance, stdX, stdY);

        Double yIntercept = this.yIntercept(slope, meanX, meanY);

        return slope != null && yIntercept != null ? slope * input - yIntercept : null;
    }

    protected Double mean(List<Double> values) {
        if (values.size() == 0) {
            return null;
        }

        double sum = values.stream().mapToDouble(Double::doubleValue).sum();

        return sum / values.size();
    }

    protected Double std(List<Double> values, Double mean) {
        if (mean == null) {
            return null;
        }

        return values.stream().reduce((double) 0, (subtotal, elt) -> subtotal + (elt - mean) * (elt - mean));
    }

    protected Double covariance(List<Double> xValues, List<Double> yValues, Double meanX, Double meanY) {
        if (xValues.size() != yValues.size() || xValues.size() == 0 || meanX == null || meanY == null) {
            return null;
        }

        double nominator = 0;

        Iterator<Double> itX = xValues.iterator();
        Iterator<Double> itY = yValues.iterator();

        while(itX.hasNext() && itY.hasNext()) {
            double x = itX.next();
            double y = itY.next();
            nominator += (x - meanX) * (y - meanY);
        }

        return nominator / xValues.size();
    }

    protected Double slope(Double covariance, Double stdX, Double stdY) {
        if (covariance == null || stdX == null || stdY == null || stdX == 0) {
            return null;
        }

        return covariance * (stdY / stdX);
    }

    protected Double yIntercept(Double slope, Double meanX, Double meanY) {
        if (slope == null || meanX == null || meanY == 0) {
            return null;
        }

        return meanY - slope * meanX;
    }

    protected List<Double> xValues(List<Double> values) {
        List<Double> xValues = new ArrayList<>();

        for (int i = 0; i < values.size(); i++) {
            xValues.add((double) i);
        }

        return xValues;
    }
}

File: generators/montithings2cpp/src/main/java/montithings/generator/steps/trafos/patterns/hwc/LinearRegression.cpp
#include "LinearRegression.h"

float LinearRegression::predict_value(float input, std::vector<float> values)
{
  if (values.empty())
  {
    return 0;
  }

  std::vector<float> x_values = this->x_values(values);

  float mean_x = this->mean(x_values);
  float mean_y = this->mean(values);

  float std_x = this->std(x_values, mean_x);
  float std_y = this->std(values, mean_y);

  float correlation = this->correlation(x_values, values, mean_x, mean_y, std_x, std_y);

  float slope = this->slope(correlation, std_x, std_y);

  float y_intercept = this->y_intercept(slope, mean_x, mean_y);

  return slope * input - y_intercept;
}

float LinearRegression::mean(std::vector<float> values)
{
  auto const count = static_cast<float>(values.size());

  return std::reduce(values.begin(), values.end()) / count;
}

float LinearRegression::std(std::vector<float> values, float mean)
{
  float subtotal = 0;
  for (int i = 0; i < values.size(); i++)
  {
    subtotal += (values[i] - mean) * (values[i] - mean);
  }

  return sqrt(1 / values.size() * subtotal);
}

float LinearRegression::correlation(std::vector<float> x_values, std::vector<float> y_values, float mean_x, float mean_y, float std_x, float std_y)
{
  if (x_values.size() != y_values.size())
  {
    throw std::invalid_argument("x values and y values don't have equal length.");
  }

  float subtotal = 0;

  for (int i = 0; i < x_values.size(); i++)
  {
    subtotal += (x_values[i] - mean_x) * (y_values[i] - mean_y);
  }

  return subtotal / std_x * std_y;
}

float LinearRegression::slope(float correlation, float std_x, float std_y)
{
  return correlation * (std_y / std_x);
}

float LinearRegression::y_intercept(float slope, float mean_x, float mean_y)
{
  return mean_y - slope * mean_x;
}

std::vector<float> LinearRegression::x_values(std::vector<float> values)
{
  std::vector<float> x_values(values.size());

  for (int i = 0; i < values.size(); i++)
  {
    x_values.push_back(i);
  }

  return x_values;
}